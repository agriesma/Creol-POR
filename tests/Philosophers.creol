interface Phil
begin
  op borrowStick
  op returnStick
end

interface Butler
begin
  with Phil op getNeighbor(out neighbor: Phil)
end

class Philosopher(butler: Butler) implements Phil 
begin 
  var hungry: Bool
  var chopstick: Bool
  var neighbor: Phil
  var history: String
 
  op init == history := ""; chopstick := true; hungry := false;
    butler.getNeighbor(; neighbor);
 
  op run == !think(); ||| !eat(); ||| !digest(); 

  op digest == await wait ; hungry := true; history := plus(history, "d");
    await wait; !digest();

  op eat == await hungry; l!neighbor.borrowStick(); await l?; await chopstick;
    history := plus(history, "e"); hungry := false; !neighbor.returnStick();
    await wait; !eat();

  op think == await not hungry; history := plus(history, "t");
    await wait; !think();

  with Phil 
    op borrowStick == await chopstick; chopstick := false;
    op returnStick == chopstick := true;
 
end

class Butler implements Butler
begin
  var p1: Phil
  var p2: Phil
  var p3: Phil
  var p4: Phil
  var p5: Phil

  op init ==
    skip;

  op run ==
    p1 := new Philosopher(this);
    p2 := new Philosopher(this);
    p3 := new Philosopher(this);
    p4 := new Philosopher(this);
    p5 := new Philosopher(this);

  with Phil
    op getNeighbor(out neighbor: Phil) ==
      if caller = p1 then neighbor := p2; else
      if caller = p2 then neighbor := p3; else
      if caller = p3 then neighbor := p4; else
      if caller = p4 then neighbor := p5; else neighbor := p1; fi fi fi fi

end

