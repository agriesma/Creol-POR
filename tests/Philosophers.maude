load interpreter
mod PROGRAM is
pr INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Philosopher" : Cl | Inh: noInh, Par: "butler", Att: "butler" |-> null , "hungry" |-> null , "chopstick" |-> null , "neighbor" |-> null , "history" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: "history" ::= str("") ; "chopstick" ::= bool(true) ; "hungry" ::= bool(false) ; "butler" . "getNeighbor" ( emp ; "neighbor" ) ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: ( "Dummy" ! "this" . "think" ( emp ) ||| "Dummy" ! "this" . "eat" ( emp ) ||| "Dummy" ! "this" . "digest" ( emp ) ) ; return ( emp ) > *
  < "digest" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await wait ) ; "hungry" ::= bool(true) ; "history" ::= "plus" ( "history" # str("d") ) ; ( await wait ) ; "Dummy" ! "this" . "digest" ( emp ) ; return ( emp ) > *
  < "eat" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await "hungry" ) ; "l" ! "neighbor" . "borrowStick" ( emp ) ; ( await ( "l" ?? ) ) ; ( await "chopstick" ) ; "history" ::= "plus" ( "history" # str("e") ) ; "hungry" ::= bool(false) ; "Dummy" ! "neighbor" . "returnStick" ( emp ) ; ( await wait ) ; "Dummy" ! "this" . "eat" ( emp ) ; return ( emp ) > *
  < "think" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await "not" ( "hungry" ) ) ; "history" ::= "plus" ( "history" # str("t") ) ; ( await wait ) ; "Dummy" ! "this" . "think" ( emp ) ; return ( emp ) > *
  < "borrowStick" : Mtdname | Param: noAid, Latt: noSubst, Code: ( await "chopstick" ) ; "chopstick" ::= bool(false) ; return ( emp ) > *
  < "returnStick" : Mtdname | Param: noAid, Latt: noSubst, Code: "chopstick" ::= bool(true) ; return ( emp ) >, Ocnt: 0 >

< "Butler" : Cl | Inh: noInh, Par: noAid, Att: "p1" |-> null , "p2" |-> null , "p3" |-> null , "p4" |-> null , "p5" |-> null, Mtds: 
  < "init" : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noAid, Latt: noSubst, Code: "p1" ::= new "Philosopher" ( "this" ) ; "p2" ::= new "Philosopher" ( "this" ) ; "p3" ::= new "Philosopher" ( "this" ) ; "p4" ::= new "Philosopher" ( "this" ) ; "p5" ::= new "Philosopher" ( "this" ) ; return ( emp ) > *
  < "getNeighbor" : Mtdname | Param: noAid, Latt: "neighbor" |-> null, Code: if "equal" ( "caller" # "p1" ) th "neighbor" ::= "p2" el if "equal" ( "caller" # "p2" ) th "neighbor" ::= "p3" el if "equal" ( "caller" # "p3" ) th "neighbor" ::= "p4" el if "equal" ( "caller" # "p4" ) th "neighbor" ::= "p5" el "neighbor" ::= "p1" fi fi fi fi ; return ( "neighbor" ) >, Ocnt: 0 >

.
endm
