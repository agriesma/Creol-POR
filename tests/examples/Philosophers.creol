/*
 * Philosophers.creol -- The dining philosophers example.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 *
 * Adapted from:
 *
 * @InProceedings{johnsen03nik,
 *   author =	 {Einar Broch Johnsen and Olaf Owe and Marte Arnestad},
 *   title =	 {Combining Active and Reactive Behavior in Concurrent
 *                Objects},
 *   editor =    {Dag Langmyhr},
 *   booktitle = {Proc. of the Norwegian Informatics Conference (NIK'03)},
 *   year =	 2003,
 *   month =	 nov,
 *   pages =     {193--204},
 *   publisher = {Tapir Academic Publisher}
 * }
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
interface Phil
begin
  with Phil
  op borrowStick
  op returnStick
end

interface Butler
begin
  with Phil op getNeighbor(out neighbor: Phil)
end

class Philosopher(butler: Butler) contracts Phil 
begin 
  var hungry: Bool
  var chopstick: Bool
  var neighbor: Phil
  var hist: String
 
  op init == hist := ""; chopstick := true; hungry := false;
    butler.getNeighbor(; neighbor)
 
  op run == !think() ; !eat() ; !digest()

  op digest == release ; hungry := true; hist := hist |- "d";
    release; !digest()

  op eat == var l: Label[ ]; await hungry; l!neighbor.borrowStick();
    await l?; await chopstick;
    hist := hist |- "e"; hungry := false; !neighbor.returnStick();
    release; !eat()

  op think == await ~ hungry; hist := hist |- "t";
    release; !think()

  with Phil 
    op borrowStick ensures ~chopstick == await chopstick; chopstick := false
    op returnStick requires ~chopstick == chopstick := true
 
end

class Butler implements Butler
begin
  var p1: Phil
  var p2: Phil
  var p3: Phil
  var p4: Phil
  var p5: Phil

  op init ==
    skip

  op run ==
    p1 := new Philosopher(this);
    p2 := new Philosopher(this);
    p3 := new Philosopher(this);
    p4 := new Philosopher(this);
    p5 := new Philosopher(this)

  with Phil
    op getNeighbor(out neighbor: Phil) ==
      if caller = p1 then neighbor := p2 else
      if caller = p2 then neighbor := p3 else
      if caller = p3 then neighbor := p4 else
      if caller = p4 then neighbor := p5 else neighbor := p1 end end end end

end

