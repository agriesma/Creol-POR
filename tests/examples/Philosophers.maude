load creol-interpreter
mod PROGRAM is
protecting CREOL-SIMULATOR .
op classes : -> Configuration [ctor] .
eq classes =
< "Philosopher" : Cl | Inh: noInh, Par: "butler", Att: "butler" |-> null , "hungry" |-> null , "chopstick" |-> null , "neighbor" |-> null , "hist" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "_" |-> null, Code: "hist" ::= str("") ; "chopstick" ::= bool(true) ; "hungry" ::= bool(false) ; "label:0" ! "butler" . "getNeighbor" ( emp ) ; ( "label:0" ? ( "neighbor" ) )  ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:2" |-> null , "label:1" |-> null , "label:0" |-> null , "_" |-> null, Code: ( "label:0" ! "this" . "think" (emp ) )  ; free( "label:0" ) ; ( "label:1" ! "this" . "eat" (emp ) )  ; free( "label:1" ) ; ( "label:2" ! "this" . "digest" (emp ) )  ; free( "label:2" ) ; return ( emp ) > *
  < "digest" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "_" |-> null, Code: release ; "hungry" ::= bool(true) ; "hist" ::= "|-" ( "hist" :: str("d") ) ; release ; ( "label:0" ! "this" . "digest" (emp ) )  ; free( "label:0" ) ; return ( emp ) > *
  < "eat" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null , "l" |-> null , "_" |-> null, Code: ( await "hungry" ) ; "l" ! "neighbor" . "borrowStick" ( emp ) ; ( await ( "l" ?? ) ) ; free( "l" ) ; ( await "chopstick" ) ; "hist" ::= "|-" ( "hist" :: str("e") ) ; "hungry" ::= bool(false) ; "label:0" ! "neighbor" . "returnStick" ( emp ) ; free( "label:0" ) ; release ; ( "label:1" ! "this" . "eat" (emp ) )  ; free( "label:1" ) ; return ( emp ) > *
  < "think" : Mtdname | Param: noVid, Latt: "label:0" |-> null , "_" |-> null, Code: ( await "~" ( "hungry" ) ) ; "hist" ::= "|-" ( "hist" :: str("t") ) ; release ; ( "label:0" ! "this" . "think" (emp ) )  ; free( "label:0" ) ; return ( emp ) > *
  < "borrowStick" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: ( await "chopstick" ) ; "chopstick" ::= bool(false) ; return ( emp ) > *
  < "returnStick" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: "chopstick" ::= bool(true) ; return ( emp ) >, Ocnt: 0 >

< "Butler" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null , "p2" |-> null , "p3" |-> null , "p4" |-> null , "p5" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: "p1" ::= new "Philosopher" ( "this" ) ; "p2" ::= new "Philosopher" ( "this" ) ; "p3" ::= new "Philosopher" ( "this" ) ; "p4" ::= new "Philosopher" ( "this" ) ; "p5" ::= new "Philosopher" ( "this" ) ; return ( emp ) > *
  < "getNeighbor" : Mtdname | Param: noVid, Latt: "neighbor" |-> null , "_" |-> null, Code: if "=" ( "caller" :: "p1" ) th "neighbor" ::= "p2" el if "=" ( "caller" :: "p2" ) th "neighbor" ::= "p3" el if "=" ( "caller" :: "p3" ) th "neighbor" ::= "p4" el if "=" ( "caller" :: "p4" ) th "neighbor" ::= "p5" el "neighbor" ::= "p1" fi fi fi fi ; return ( "neighbor" ) >, Ocnt: 0 >

.
endm
