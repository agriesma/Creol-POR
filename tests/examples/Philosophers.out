==========================================
reduce in PROGRAM : init .
result Configuration: 
< "Butler" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null, "p2" |-> null,
    "p3" |-> null, "p4" |-> null, "p5" |-> null, Mtds:
  < "getNeighbor" : Mtdname | Param: noVid, Latt: "neighbor" |-> null, Code: if
    "="("caller" # "p1") th "neighbor" ::= "p2" el if "="("caller" # "p2") th
    "neighbor" ::= "p3" el if "="("caller" # "p3") th "neighbor" ::= "p4" el if
    "="("caller" # "p4") th "neighbor" ::= "p5" el "neighbor" ::= "p1" fi fi fi
    fi ; return("neighbor") > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >
    *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new
    "Philosopher"("this") ; "p2" ::= new "Philosopher"("this") ; "p3" ::= new
    "Philosopher"("this") ; "p4" ::= new "Philosopher"("this") ; "p5" ::= new
    "Philosopher"("this") ; return(emp) >,
  Ocnt: 0 >

< "Philosopher" : Cl | Inh: noInh, Par: "butler", Att: "butler" |-> null,
    "chopstick" |-> null, "hist" |-> null, "hungry" |-> null, "neighbor" |->
    null, Mtds:
  < "borrowStick" : Mtdname | Param: noVid, Latt: noSubst, Code: await
    "chopstick" ; "chopstick" ::= bool(false) ; return(emp) > *
  < "digest" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code: release
    ; "hungry" ::= bool(true) ; "hist" ::= "|-"("hist" # str("d")) ; release ;
    "label:0" ! "this" . "digest"(emp) ; free("label:0") ; return(emp) > *
  < "eat" : Mtdname | Param: noVid, Latt: "l" |-> null, "label:0" |-> null,
    "label:1" |-> null, Code: await "hungry" ; "l" ! "neighbor" .
    "borrowStick"(emp) ; await ("l" ??) ; await "chopstick" ; "hist" ::= "|-"(
    "hist" # str("e")) ; "hungry" ::= bool(false) ; "label:0" ! "neighbor" .
    "returnStick"(emp) ; free("label:0") ; release ; "label:1" ! "this" .
    "eat"(emp) ; free("label:1") ; return(emp) > *
  < "init" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code: "hist" ::=
    str("") ; "chopstick" ::= bool(true) ; "hungry" ::= bool(false) ; "label:0"
    ! "butler" . "getNeighbor"(emp) ; "label:0" ?("neighbor") ; return(emp) > *
  < "returnStick" : Mtdname | Param: noVid, Latt: noSubst, Code: "chopstick"
    ::= bool(true) ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, "label:2" |-> null, Code: ("label:0" ! "this" . "think"(emp) ; free(
    "label:0") ||| "label:1" ! "this" . "eat"(emp) ; free("label:1") |||
    "label:2" ! "this" . "digest"(emp) ; free("label:2")) ; return(emp) > *
  < "think" : Mtdname | Param: noVid, Latt: "label:0" |-> null, Code: await
    "~"("hungry") ; "hist" ::= "|-"("hist" # str("t")) ; release ; "label:0" !
    "this" . "think"(emp) ; free("label:0") ; return(emp) >,
  Ocnt: 0 >
Bye.
